casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -79,32 +84,50 @@ sealed abstract class MultiParentCasperInstances {
     }
 
   def hashSetCasper[
-      F[_]: Monad: Capture: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler: SafetyOracle](
+      F[_]: Monad: Capture: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler: SafetyOracle: BlockStore](
       runtimeManager: RuntimeManager,
       validatorId: Option[ValidatorIdentity],
-      genesis: BlockMessage)(implicit scheduler: Scheduler): MultiParentCasper[F] =
+      genesis: BlockMessage)(implicit scheduler: Scheduler): F[MultiParentCasper[F]] =
+    for {
+      _           <- BlockStore[F].put(genesis.blockHash, genesis)
+      internalMap <- BlockStore[F].asMap()
+      dag         = BlockDag()
+      (maybePostGenesisStateHash, _) = InterpreterUtil
+        .validateBlockCheckpoint(
+          genesis,
+          genesis,
+          dag,
+          internalMap,
+          runtimeManager.emptyStateHash,
+          Set[StateHash](runtimeManager.emptyStateHash),
+          runtimeManager
+        )
+    } yield {
+      createMultiParentCasper[F](runtimeManager,
+                                 validatorId,
+                                 genesis,
+                                 dag,
+                                 maybePostGenesisStateHash)
+    }
+
+  private[this] def createMultiParentCasper[
+      F[_]: Monad: Capture: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler: SafetyOracle: BlockStore](
+      runtimeManager: RuntimeManager,
+      validatorId: Option[ValidatorIdentity],
+      genesis: BlockMessage,
+      dag: BlockDag,
+      maybePostGenesisStateHash: Option[StateHash])(implicit scheduler: Scheduler) =
     new MultiParentCasper[F] {
       type BlockHash = ByteString
       type Validator = ByteString
 
       //TODO: Extract hardcoded version
       private val version = 0L
 
-      private val _blockDag: AtomicSyncVar[BlockDag] = new AtomicSyncVar(
-        BlockDag().copy(
-          blockLookup = HashMap[BlockHash, BlockMessage](genesis.blockHash -> genesis))
-      )
+      private val _blockDag: AtomicSyncVar[BlockDag] = new AtomicSyncVar(dag)
+
       private val emptyStateHash = runtimeManager.emptyStateHash
 
-      private val (maybePostGenesisStateHash, _) = InterpreterUtil
-        .validateBlockCheckpoint(
-          genesis,
-          genesis,
-          _blockDag.get,
-          emptyStateHash,
-          Set[StateHash](emptyStateHash),
-          runtimeManager
-        )
```

Comment:
We need to reconstruct the knownStateHashesContainer below with something like

`val knownStateHashesContainer: AtomicSyncVar[Set[StateHash]] = new AtomicSyncVar(BlockStore.asMap().map(block => ProtoUtil.tuplespace(block)))`